File: src/main/java/com/chessonline/Knight.java
package com.chessonline;

public class Knight extends Piece {

    public Knight(String color, String position) {
        super("knight", color, position);
    }

    @Override
    public boolean move(String newPosition) {

        if (!isMovementValid(newPosition)) return false;

        setPosition(newPosition);

        return true;
    }

    @Override
    public boolean isMovementValid(String newPosition) {
        Position current = new Position(position);
        Position target = new Position(newPosition);

        int colDiff = Math.abs(target.col - current.col);
        int rowDiff = Math.abs(target.row - current.row);

        // Mouvement en "L" : 2 cases dans une direction, 1 case perpendiculaire
        return (colDiff == 2 && rowDiff == 1) || (colDiff == 1 && rowDiff == 2);
    }

    @Override
    public boolean isCaptureMovementValid(String newPosition) {
        return isMovementValid(newPosition);
    }
}

File: engine/pieces/knight.py
from .piece import Piece

class Knight(Piece):
    def __init__(self, color: str, position: str):
        super().__init__("knight", color, position)

    def move(self, new_position: str) -> bool:
        if not self.is_movement_valid(new_position):
            return False

        self.set_position(new_position)
        return True

    def is_movement_valid(self, new_position: str) -> bool:
        current_col = self.position[0]
        current_row = int(self.position[1])

        target_col = new_position[0]
        target_row = int(new_position[1])

        col_diff = abs(ord(target_col) - ord(current_col))
        row_diff = abs(target_row - current_row)

        # L-shape movement
        return (col_diff == 2 and row_diff == 1) or (col_diff == 1 and row_diff == 2)

    def is_capture_movement_valid(self, new_position: str) -> bool:
        return self.is_movement_valid(new_position)